### PR TITLE
Use Indexer for inserts on primary

### PR DESCRIPTION
**Pull request by mfussenegger**
First step to actually use the Indexer. So far limited to the insert operation on the primary.

This also contains a couple of more fixes around the indexers which were uncovered by integration tests

### CODE REVIEW

File: server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java [hunk 1]
```diff
@@ -319,12 +351,81 @@ static class IndexItemResponse {
...
+
+        if (!parsedDoc.newColumns().isEmpty()) {
+            var addColumnRequest = new AddColumnRequest(
+                RelationName.fromIndexName(indexShard.shardId().getIndexName()),
+                parsedDoc.newColumns(),
+                Map.of(),
+                new IntArrayList(0)
+            );
+            addColumnAction.execute(addColumnRequest).get();
+        }
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java:383]: Do I get it right, that old mapping update via TransportSchardUpdteAxction called in the parent abstract class 
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/dml/TransportShardAction.java#L96
is not used in the indexer but only in `legacyInsert` and will be gone eventually?

**Review by BaurzhanSakhariev — APPROVED** [review]
Added a question, LGTM otherwise

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -770,7 +770,23 @@ public static Engine.Index prepareIndex(DocumentMapper docMapper,
                                 ifSeqNo, ifPrimaryTerm);
     }
 
-    private Engine.IndexResult index(Engine engine, Engine.Index index) throws IOException {
+    public Engine.IndexResult index(Engine.Index index) throws IOException {
+        assert index.primaryTerm() <= getOperationPrimaryTerm()
+                : "op term [ " + index.primaryTerm() + " ] > shard term [" + getOperationPrimaryTerm() + "]";
+        ensureWriteAllowed(index.origin());
+        try {
+            return index(getEngine(), index);
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:778]: very minor: can't we skip the `Engine` arg and call `getEngine` in the `index()` method?

File: server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java [hunk 3]
```diff
@@ -319,12 +351,81 @@ static class IndexItemResponse {
...
+        item.source(parsedDoc.source());
+
+        if (!parsedDoc.newColumns().isEmpty()) {
+            var addColumnRequest = new AddColumnRequest(
+                RelationName.fromIndexName(indexShard.shardId().getIndexName()),
+                parsedDoc.newColumns(),
+                Map.of(),
+                new IntArrayList(0)
+            );
+            addColumnAction.execute(addColumnRequest).get();
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java:382]: is it ok to call `get()` and make it sync here?

**Review by matriv — APPROVED** [review]
LGTM. Left a minor comment and a question.

File: server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java (hunk 1 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java:383]: Yes

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java (hunk 2 above)
Comment (mfussenegger) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java:778]: The method this is calling into is currently also called from other places where the engine is retrieved earlier.

File: server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java (hunk 3 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/dml/upsert/TransportShardUpsertAction.java:382]: It's not good but matches the previous behavior. Making the whole process async is a bigger topic.

**Pull request merged**